Windows UIA: 0x80040201 (EVENT_E_ALL_SUBSCRIBERS_FAILED) still fatal on query path (wait_visible/exists) — follow-up to #169

## Summary

`0x80040201` (`EVENT_E_ALL_SUBSCRIBERS_FAILED`) still surfaces as a fatal `xa11y.PlatformError` on Windows UIA — but now from the **query / element-discovery path** (e.g. `wait_visible()`, `exists()`, tree snapshot), not from `press()`. This is the same transient Qt-UIA-backend failure that #169 fixed for **action** methods, but the fix was never extended to the read path.

This is a **follow-up / partial regression of #169** (which is closed). #169 added the `is_event_subscriber_failure(&e) → Ok(())` guard to `press()` / `toggle()` / `select()`, and that part works. But callers that *query* the tree still get the raw error.

## Environment

- **xa11y: 0.8.2** (`xa11y >= 0.8.2, < 0.9`, PyPI `cp39-abi3-win_amd64` wheel)
- OS: **Windows GitHub-hosted runner** (`windows-latest`)
- Python: 3.13
- App under test: PySide6 / Qt 6 launched as a subprocess (`deadline bundle gui-submit …` / `deadline config gui`)
- Backend: UIA

## Evidence — real CI runs

These are intermittent: the same commit/test passes on re-run. All are on Windows; the failing operation is always a **read**, not a `.press()`. (GitHub retains raw logs ~90 days, so the key excerpt is inlined under each link.)

**1. `mainline` push — `wait_visible()` during launch**
https://github.com/aws-deadline/deadline-cloud/actions/runs/27177528841/job/80229545390
```
test/ui/test_bundle_gui_submit_json.py::TestSubmitterName::test_submit_exits_process
  test\ui\helpers.py:431: in open
  test\ui\helpers.py:250: in launch
  test\ui\helpers.py:297: in _launch_once
      instance.dialog().wait_visible(timeout=timeout)   # <-- query, not an action
  E   xa11y.PlatformError: Platform error (-2147220991): An event was unable to invoke any of the subscribers (0x80040201)
```

**2. `mainline` push — `TestSubmitterOpens::test_has_tabs`**
https://github.com/aws-deadline/deadline-cloud/actions/runs/26906463874/job/79372581326

**3. dependabot PR (unrelated change) — `test_job_name_displayed`**
https://github.com/aws-deadline/deadline-cloud/actions/runs/26976111323/job/79603417579

**4. PR — `TestConfigGuiOpens::test_dialog_is_visible`**
https://github.com/aws-deadline/deadline-cloud/actions/runs/26663966584/job/78593009624

Over the last month this signature accounted for **~12 distinct failing CI runs**, second only to bridge-startup timeouts. Affected tests are simply whichever run first / query the tree earliest (`TestSubmitterOpens`, `TestConfigGuiOpens`, `test_priority_spin_box_is_present`); it is not test-specific.

## Root cause (code pointer)

`xa11y-windows/src/uia.rs` already classifies this HRESULT as transient and benign:

```rust
// xa11y-windows/src/uia.rs:21
/// EVENT_E_ALL_SUBSCRIBERS_FAILED (0x80040201) — ... the UI action itself
/// completed; only the notification layer had a transient failure. Certain
/// providers (notably Qt's UIA backend) propagate this error back ... which
/// is incorrect — callers should treat it as success. See: #169
```

The `press()` / `toggle()` / `select()` paths handle it:

```rust
unsafe { pattern.Invoke() }.or_else(|e| {
    if is_event_subscriber_failure(&e) { Ok(()) }
    else { Err(Error::Platform { ... }) }
})?;
```

But the **discovery / snapshot path does not**. `find_all_subtree()` (which backs `exists()`, `.element()`, `wait_visible()`, `dump()`) routes through the generic helper, which maps every error — including `0x80040201` — straight to a fatal `Error::Platform`:

```rust
// xa11y-windows/src/uia.rs
fn uia_call<T>(f: impl FnOnce() -> windows::core::Result<T>) -> Result<T> {
    f().map_err(|e| Error::Platform {           // <-- no is_event_subscriber_failure check
        code: e.code().0 as i64,
        message: e.to_string(),
    })
}
```

## Suggested fix

Treat `EVENT_E_ALL_SUBSCRIBERS_FAILED` as transient on the read path too. Since a read needs a *value* (not just "success"), swallowing isn't enough — a short retry is the natural fit, e.g. in `uia_call` or at the `FindAllBuildCache` sites:

```rust
fn uia_call<T>(f: impl Fn() -> windows::core::Result<T>) -> Result<T> {
    for attempt in 0..3 {
        match f() {
            Ok(v) => return Ok(v),
            Err(e) if is_event_subscriber_failure(&e) && attempt < 2 => {
                std::thread::sleep(Duration::from_millis(50));
                continue;
            }
            Err(e) => return Err(Error::Platform { code: e.code().0 as i64, message: e.to_string() }),
        }
    }
    unreachable!()
}
```

(Exact shape is yours to decide — retry-in-`uia_call`, or a dedicated retry wrapper only around the subtree fetch so unrelated calls aren't slowed.) The key point is that the query path should be as tolerant of `0x80040201` as the action path already is — the original issue's rationale ("the action itself completed; only the notification layer hiccupped") applies equally to a tree read.

Happy to test a pre-release wheel against our Windows CI matrix if that helps validate.

--- a/docs/site/src/content/docs/guides/accessibility-quirks.mdx
+++ b/docs/site/src/content/docs/guides/accessibility-quirks.mdx
@@ -282,7 +282,9 @@ EVENT_E_ALL_SUBSCRIBERS_FAILED (0x80040201)
 
 Qt propagates this back through the action call **even though the UI action
 itself completed successfully**. The same code can surface from
-`FindAllBuildCache` during a find. Treat `0x80040201` as success.
+`FindAllBuildCache` during a find. xa11y treats `0x80040201` as success on
+action calls (`press`/`toggle`/`select`), and on query calls — where a value is
+still needed — retries the call a few times before giving up.
 
 ### Windows: `FindAllBuildCache` silently drops fragment elements
 

--- a/xa11y-windows/src/uia.rs
+++ b/xa11y-windows/src/uia.rs
@@ -274,12 +274,40 @@ impl WindowsProvider {
 
 // ── Safe UIA helpers ────────────────────────────────────────────────────────
 
+/// Maximum number of attempts for a UIA call that keeps failing with
+/// `EVENT_E_ALL_SUBSCRIBERS_FAILED`, and the delay between attempts.
+const EVENT_SUBSCRIBER_FAILURE_ATTEMPTS: u32 = 3;
+const EVENT_SUBSCRIBER_FAILURE_RETRY_DELAY: std::time::Duration =
+    std::time::Duration::from_millis(50);
+
 /// Wrap a UIA COM call, mapping the error to xa11y Error::Platform.
-fn uia_call<T>(f: impl FnOnce() -> windows::core::Result<T>) -> Result<T> {
-    f().map_err(|e| Error::Platform {
-        code: e.code().0 as i64,
-        message: e.to_string(),
-    })
+///
+/// `EVENT_E_ALL_SUBSCRIBERS_FAILED` (0x80040201) is transient (see the
+/// constant's doc above): some providers (notably Qt's UIA backend) surface
+/// it from query calls like `FindAllBuildCache` even though only the
+/// notification layer hiccupped. The action paths (`press`/`toggle`/`select`)
+/// can swallow it outright because the action already completed (#169); a
+/// query needs a value, so the call is retried a few times before the error
+/// is propagated. Any other error is returned immediately — this is a retry
+/// of one classified-transient HRESULT, not a fallback (tenet 1).
+/// See: https://github.com/xa11y/xa11y/issues/257
+fn uia_call<T>(f: impl Fn() -> windows::core::Result<T>) -> Result<T> {
+    let mut attempts_left = EVENT_SUBSCRIBER_FAILURE_ATTEMPTS;
+    loop {
+        attempts_left -= 1;
+        match f() {
+            Ok(v) => return Ok(v),
+            Err(e) if is_event_subscriber_failure(&e) && attempts_left > 0 => {
+                std::thread::sleep(EVENT_SUBSCRIBER_FAILURE_RETRY_DELAY);
+            }
+            Err(e) => {
+                return Err(Error::Platform {
+                    code: e.code().0 as i64,
+                    message: e.to_string(),
+                })
+            }
+        }
+    }
 }
 
 /// Read a BSTR VARIANT property from the element's pre-fetched snapshot.
@@ -2562,6 +2590,78 @@ mod tests {
                 !is_event_subscriber_failure(&err),
                 "HRESULT 0x{code:08X} must not be classified as an event-subscriber failure"
             );
+        }
+    }
+
+    // ── uia_call retry behaviour (issue #257) ───────────────────────────────
+
+    fn subscriber_failure() -> windows::core::Error {
+        EVENT_E_ALL_SUBSCRIBERS_FAILED.ok().unwrap_err()
+    }
+
+    #[test]
+    fn uia_call_success_calls_once() {
+        let calls = std::cell::Cell::new(0u32);
+        let result = uia_call(|| {
+            calls.set(calls.get() + 1);
+            Ok(42)
+        });
+        assert_eq!(result.unwrap(), 42);
+        assert_eq!(calls.get(), 1);
+    }
+
+    #[test]
+    fn uia_call_retries_event_subscriber_failure_then_succeeds() {
+        // Two transient 0x80040201 failures, then success — the query path
+        // must ride out the same Qt-UIA hiccup the action path tolerates.
+        let calls = std::cell::Cell::new(0u32);
+        let result = uia_call(|| {
+            calls.set(calls.get() + 1);
+            if calls.get() < EVENT_SUBSCRIBER_FAILURE_ATTEMPTS {
+                Err(subscriber_failure())
+            } else {
+                Ok("tree")
+            }
+        });
+        assert_eq!(result.unwrap(), "tree");
+        assert_eq!(calls.get(), EVENT_SUBSCRIBER_FAILURE_ATTEMPTS);
+    }
+
+    #[test]
+    fn uia_call_propagates_persistent_event_subscriber_failure() {
+        // If every attempt fails with 0x80040201, the error must still
+        // surface (no infinite retry, no silent swallow on the read path —
+        // unlike an action, a query has no value to return).
+        let calls = std::cell::Cell::new(0u32);
+        let result: Result<()> = uia_call(|| {
+            calls.set(calls.get() + 1);
+            Err(subscriber_failure())
+        });
+        assert_eq!(calls.get(), EVENT_SUBSCRIBER_FAILURE_ATTEMPTS);
+        match result {
+            Err(Error::Platform { code, .. }) => {
+                assert_eq!(code, EVENT_E_ALL_SUBSCRIBERS_FAILED.0 as i64);
+            }
+            other => panic!("expected Error::Platform, got {other:?}"),
+        }
+    }
+
+    #[test]
+    fn uia_call_does_not_retry_other_errors() {
+        let calls = std::cell::Cell::new(0u32);
+        let e_fail = windows::core::HRESULT(0x80004005u32 as i32);
+        let result: Result<()> = uia_call(|| {
+            calls.set(calls.get() + 1);
+            Err(e_fail.ok().unwrap_err())
+        });
+        assert_eq!(
+            calls.get(),
+            1,
+            "non-transient errors must fail on the first attempt"
+        );
+        match result {
+            Err(Error::Platform { code, .. }) => assert_eq!(code, e_fail.0 as i64),
+            other => panic!("expected Error::Platform, got {other:?}"),
         }
     }
 }

